src/IECoreArnold/ShaderNetworkAlgo.cpp

```diff
@@ -198,65 +198,73 @@ AtNode *convertWalk( const ShaderNetwork::Parameter &outputParameter, const IECo
 		return node;
 	}
 
-	// Set the shader parameters
+	// Recurse through input connections
+	std::set< InternedString > connected;
 
-	for( const auto &namedParameter : shader->parameters() )
+	for( const auto &connection : shaderNetwork->inputConnections( outputParameter.shader ) )
 	{
+		AtNode *sourceNode = convertWalk( connection.source, shaderNetwork, name, nodeCreator, nodes, converted );
+		if( !sourceNode )
+		{
+			continue;
+		}
+
 		string parameterName;
 		if( isOSLShader )
 		{
-			parameterName = "param_" + namedParameter.first.string();
+			parameterName = "param_" + connection.destination.name.string();
 		}
 		else
 		{
-			parameterName = namedParameter.first.string();
+			parameterName = connection.destination.name.string();
 		}
 
-		if( const SplineffData *splineData = runTimeCast<const SplineffData>( namedParameter.second.get() ) )
-		{
-			setSplineParameter( node, parameterName, splineData->readable() );
-		}
-		else if( const SplinefColor3fData *splineData = runTimeCast<const SplinefColor3fData>( namedParameter.second.get() ) )
-		{
-			setSplineParameter( node, parameterName, splineData->readable() );
-			continue;
-		}
-		else
+		InternedString sourceName = connection.source.name;
+		const IECoreScene::Shader *sourceShader = shaderNetwork->getShader( connection.source.shader );
+		if( boost::starts_with( sourceShader->getType(), "osl:" ) )
 		{
-			ParameterAlgo::setParameter( node, AtString( parameterName.c_str() ), namedParameter.second.get() );
+			sourceName = partitionEnd( sourceName, '.' );
 		}
+
+		AiNodeLinkOutput( sourceNode, sourceName.c_str(), node, parameterName.c_str() );
+		connected.insert( connection.destination.name );
 	}
 
-	// Recurse through input connections
+	// Set the shader parameters
 
-	for( const auto &connection : shaderNetwork->inputConnections( outputParameter.shader ) )
+	for( const auto &namedParameter : shader->parameters() )
 	{
-		AtNode *sourceNode = convertWalk( connection.source, shaderNetwork, name, nodeCreator, nodes, converted );
-		if( !sourceNode )
+		if( connected.count( namedParameter.first ) )
 		{
+			// Don't set values on parameters that are connected
```

Comment:
This rearrangement of the logic is a bit misleading I think - we're still setting values on parameters that have connections at the component (`color.r`) level. This is completely necessary, and was more obviously intentional with the previous logic. If we can make it work, I definitely prefer the approach of just transferring all the values we have, and all the connections.

Reply:
I've confirmed that the issue seems to be particular to quad_light.color.  We could do a specific fix, but your proposed solution of calling reset when we connect an input is easy enough that maybe there's no harm in doing it unconditionally.  I've done that here: 158da5e